requestmanager/asyncloader/asyncloader.go

```diff
@@ -71,170 +65,112 @@ func New(ctx context.Context, linkSystem ipld.LinkSystem, allocator Allocator) *
+	delete(al.alternateQueues, name)
+	return nil
 }
 
 // StartRequest indicates the given request has started and the manager should
 // continually attempt to load links for this request as new responses come in
 func (al *AsyncLoader) StartRequest(requestID graphsync.RequestID, persistenceOption string) error {
-	response := make(chan error, 1)
-	err := al.sendSyncMessage(&startRequestMessage{requestID, persistenceOption, response}, response)
-	return err
+	al.stateLk.Lock()
+	defer al.stateLk.Unlock()
+	if persistenceOption != "" {
+		_, ok := al.alternateQueues[persistenceOption]
+		if !ok {
+			return errors.New("unknown persistence option")
+		}
+		al.requestQueues[requestID] = persistenceOption
+	}
+	al.activeRequests[requestID] = struct{}{}
+	return nil
 }
 
 // ProcessResponse injests new responses and completes asynchronous loads as
 // neccesary
 func (al *AsyncLoader) ProcessResponse(responses map[graphsync.RequestID]metadata.Metadata,
 	blks []blocks.Block) {
-	select {
-	case <-al.ctx.Done():
-	case al.incomingMessages <- &newResponsesAvailableMessage{responses, blks}:
+	al.stateLk.Lock()
+	defer al.stateLk.Unlock()
+	byQueue := make(map[string][]graphsync.RequestID)
+	for requestID := range responses {
+		queue := al.requestQueues[requestID]
+		byQueue[queue] = append(byQueue[queue], requestID)
+	}
+	for queue, requestIDs := range byQueue {
+		loadAttemptQueue := al.getLoadAttemptQueue(queue)
+		responseCache := al.getResponseCache(queue)
+		queueResponses := make(map[graphsync.RequestID]metadata.Metadata, len(requestIDs))
+		for _, requestID := range requestIDs {
+			queueResponses[requestID] = responses[requestID]
+		}
+		responseCache.ProcessResponse(queueResponses, blks)
+		loadAttemptQueue.RetryLoads()
 	}
 }
 
 // AsyncLoad asynchronously loads the given link for the given request ID. It returns a channel for data and a channel
 // for errors -- only one message will be sent over either.
 func (al *AsyncLoader) AsyncLoad(p peer.ID, requestID graphsync.RequestID, link ipld.Link, linkContext ipld.LinkContext) <-chan types.AsyncLoadResult {
 	resultChan := make(chan types.AsyncLoadResult, 1)
-	response := make(chan error, 1)
 	lr := loadattemptqueue.NewLoadRequest(p, requestID, link, linkContext, resultChan)
-	_ = al.sendSyncMessage(&loadRequestMessage{response, requestID, lr}, response)
+	al.stateLk.Lock()
+	defer al.stateLk.Unlock()
+	_, retry := al.activeRequests[requestID]
+	loadAttemptQueue := al.getLoadAttemptQueue(al.requestQueues[requestID])
+	loadAttemptQueue.AttemptLoad(lr, retry)
 	return resultChan
 }
 
 // CompleteResponsesFor indicates no further responses will come in for the given
 // requestID, so if no responses are in the cache or local store, a link load
 // should not retry
 func (al *AsyncLoader) CompleteResponsesFor(requestID graphsync.RequestID) {
-	select {
-	case <-al.ctx.Done():
-	case al.incomingMessages <- &finishRequestMessage{requestID}:
-	}
+	al.stateLk.Lock()
+	defer al.stateLk.Unlock()
```

Comment:
this PR adds this single mutex lock around the loader's state. Do we have reasonable confidence that we won't end up with lock contention over it?

Reply:
The old code sent a message to a queue channel, and a single goroutine owned the state and processed those messages sequentially. So the amount of contention seems basically the same before and after.

One significant difference is the priority and ordering; with a channel, before you would be guaranteed first-come-first-serve. With a mutex, it's always a semi-random race to decide what goroutine gets to grab the mutex, so if incoming messages keep coming, you could have high latencies with older messages getting unlucky and not grabbing the mutex. It's not clear to me whether order matters here.

One last thought: sharing memory does give us more ability to be performant, if we want to. For example, using a RWMutex, or multiple mutexes, or sync/atomic. So in terms of contention, I think it can be better than a single goroutine with a message queue, if we care enough to optimize it.

Reply:
Yea I feel quite confident about this being a better, more reliable strategy.